zcash_client_backend/src/data_api.rs

```diff
@@ -0,0 +1,583 @@
+
+    /// Returns the memo for a received note, if it is known and a valid UTF-8 string.
+    fn get_received_memo_as_utf8(
+        &self,
+        id_note: Self::NoteRef,
+    ) -> Result<Option<String>, Self::Error>;
+
+    /// Returns the memo for a sent note, if it is known and a valid UTF-8 string.
+    fn get_sent_memo_as_utf8(&self, id_note: Self::NoteRef) -> Result<Option<String>, Self::Error>;
+
+    /// Returns the note commitment tree at the specified block height.
+    fn get_commitment_tree(
+        &self,
+        block_height: BlockHeight,
+    ) -> Result<Option<CommitmentTree<Node>>, Self::Error>;
+
+    /// Returns the incremental witnesses as of the specified block height.
+    fn get_witnesses(
+        &self,
+        block_height: BlockHeight,
+    ) -> Result<Vec<(Self::NoteRef, IncrementalWitness<Node>)>, Self::Error>;
+
+    /// Returns the unspent nullifiers, along with the account identifiers
+    /// with which they are associated.
+    fn get_nullifiers(&self) -> Result<Vec<(Nullifier, AccountId)>, Self::Error>;
+
+    /// Return all spendable notes.
+    fn get_spendable_notes(
+        &self,
+        account: AccountId,
+        anchor_height: BlockHeight,
+    ) -> Result<Vec<SpendableNote>, Self::Error>;
+
+    /// Returns a list of spendable notes sufficient to cover the specified
+    /// target value, if possible.
+    fn select_spendable_notes(
+        &self,
+        account: AccountId,
+        target_value: Amount,
+        anchor_height: BlockHeight,
+    ) -> Result<Vec<SpendableNote>, Self::Error>;
+}
+
+/// This trait encapsulates the write capabilities required to update stored
+/// wallet data.
+pub trait WalletWrite: WalletRead {
+    /// Perform one or more write operations of this trait transactionally.
+    /// Implementations of this method must ensure that all mutations to the
+    /// state of the data store made by the provided closure must be performed
+    /// atomically and modifications to state must be automatically rolled back
+    /// if the provided closure returns an error.
+    fn transactionally<F, A>(&mut self, f: F) -> Result<A, Self::Error>
+    where
+        F: FnOnce(&mut Self) -> Result<A, Self::Error>;
+
+    /// Add the data for a block to the data store.
+    fn insert_block(
+        &mut self,
+        block_height: BlockHeight,
+        block_hash: BlockHash,
+        block_time: u32,
+        commitment_tree: &CommitmentTree<Node>,
+    ) -> Result<(), Self::Error>;
+
+    /// This method assumes that the state of the underlying data store is
+    /// consistent up to a particular block height. Since it is possible that
+    /// a chain reorg might invalidate some stored state, this method must be
+    /// implemented in order to allow users of this API to "reset" the data store
+    /// to correctly represent chainstate as of a specified block height.
+    fn rewind_to_height(&mut self, block_height: BlockHeight) -> Result<(), Self::Error>;
```

Comment:
To ward away off-by-one errors, say explicitly that after calling this method, the block at the given height will be the most recent block included in the state. Also document any restrictions on how far it's possible to rewind.